### PR TITLE
Improves boot time

### PR DESCRIPTION
**Pull request by arcanis**
**What's the problem this PR addresses?**

Yarn is currently a little slow to start. Even a `yarn --version` takes ~280ms on my (slow) machine.

**How did you fix it?**

~~This diff implements bundle splitting by adding a `--split` option to the bundle generator that, when used, causes it to emit a self-extracting file that unpacks the chunks on disk then runs them. In practice it isn't more performant, so it's unlikely to be enabled by default, but having this capability can be useful for experimentation purposes.~~

~~A few other runtime speed improvements not directly related to it are also implemented.~~

- Reimplemented Tinylogic using Pegjs rather than Chevrotine
- Changed some of the `got` import into type imports
- Removed `yup` from the modules exposed to plugins (was legacy)

```sh
YARN_IGNORE_PATH=1 hyperfine -w 1 "node ./master.cjs --version" "node ./after.cjs --version"
Benchmark 1: node ./master.cjs --version
  Time (mean ± σ):     214.8 ms ±   1.1 ms    [User: 228.5 ms, System: 18.1 ms]
  Range (min … max):   213.4 ms … 216.9 ms    13 runs

Benchmark 2: node ./after.cjs --version
  Time (mean ± σ):     178.0 ms ±   1.3 ms    [User: 189.2 ms, System: 20.0 ms]
  Range (min … max):   175.7 ms … 180.8 ms    16 runs

Summary
  'node ./after.cjs --version' ran
    1.21 ± 0.01 times faster than 'node ./master.cjs --version'
```

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

**Review by merceyz — CHANGES_REQUESTED** [review]

File: packages/yarnpkg-libzip/sources/sync.ts [hunk 1]
```diff
@@ -1,17 +1,24 @@
...
 export async function getLibzipPromise() {
-  return getLibzipSync();
+  if (asyncLibzip === null) {
+    asyncLibzip = import('./libzipAsync').then(async ({default: createModule}) => {
+      const libzip = await createModule();
+      return makeInterface(libzip);
+    });
+  }
+
+  return asyncLibzip;
```
[the remark this answers is not in the document]
Comment (arcanis) [reply packages/yarnpkg-libzip/sources/sync.ts:21]: The sync version isn't used apart from our tests, so I'd expect it to be tree-shaken; I'll double-check that it isn't included twice

File: packages/yarnpkg-builder/sources/commands/build/bundle.ts [hunk 2]
```diff
@@ -59,6 +60,10 @@ export default class BuildBundleCommand extends Command {
     description: `An array of plugins that should be included besides the ones specified in the profile`,
   });
 
+  split = Option.Boolean(`--split`, false, {
+    description: `Build a bundle that splits the application then unifies it into a self-extracting packaged app`,
+  });
```
[the remark this answers is not in the document]
Comment (arcanis) [reply packages/yarnpkg-builder/sources/commands/build/bundle.ts:65]: Fair enough, I mostly added it to enable further experimentations, but I'm fine with reverting it. Worst case it'll still be available as a standalone commit in this original PR 😄

File: packages/yarnpkg-libzip/sources/sync.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (arcanis) [reply packages/yarnpkg-libzip/sources/sync.ts]: It wasn't tree shaken; reverted this file

File: packages/yarnpkg-cli/sources/tools/backportClipanionCompatibility.ts [hunk 3]
```diff
@@ -1,24 +0,0 @@
-export function backportClipanionCompatibility(clipanion: any) {
```
Comment (merceyz) [inline packages/yarnpkg-cli/sources/tools/backportClipanionCompatibility.ts:1]: Removing this file belongs in https://github.com/yarnpkg/berry/pull/4262

**Review by merceyz — CHANGES_REQUESTED** [review]
> Moved the builtin patch behind lazy imports
Moved the worker hook behind a lazy import
Moved the PnP hook behind a lazy import

Without bundle splitting these can be reverted as they are just functions

File: packages/yarnpkg-cli/sources/tools/getPluginConfiguration.val.js [hunk 4]
```diff
@@ -7,7 +7,7 @@ module.exports = ({modules, plugins}) => {
   }).join(``);
 
   const moduleSegment = `  modules: new Map([\n${modules.map((request, index) => {
-    return `    [${JSON.stringify(require(`${request}/package.json`).name)}, ${request === `clipanion` ? `backportClipanionCompatibility` : ``}(_${index})],\n`;
+    return `    [${JSON.stringify(require(`${request}/package.json`).name)}, _${index}],\n`;
```
[the remark this answers is not in the document]
Comment (merceyz) [reply packages/yarnpkg-cli/sources/tools/getPluginConfiguration.val.js]: Ref https://github.com/yarnpkg/berry/pull/4262

**Review by merceyz — APPROVED** [review]

**Review by merceyz — APPROVED** [review]

**Pull request merged**